### PR TITLE
[VKCI-295] Release IP on load balancer Deletion

### PR DESCRIPTION
**Pull request by rocknes**
Added method:
ReleaseIpFromLoadBalancer in gateway.go

Updated method:
DeleteLoadBalancer in gateway.go

Testing Done:
Add testcase TestIpAllocationAndReleaseToLoadBalancer to ip_spaces_test.go
Fixed broken tests in gateway_system_test.go

- - -
This change is [Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/349)

### CODE REVIEW

**Review by lzichong — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/349#-:-NqZ0T_MCq9UwJQrZglf:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @Anirudh9794, @arunmk, and @sahithi)

File: pkg/vcdsdk/gateway.go [hunk 1]
```diff
@@ -2209,3 +2224,51 @@ func (gm *GatewayManager) reserveIpForLoadBalancer(ctx context.Context, claimMar
 	// Was unable to reserve an Ip on any of the available Ip Spaces
 	return "", fmt.Errorf("unable to reserve Ip from any available Ip spaces")
 }
+
+// releaseIpFromLoadBalancer will scan through all Ip Spaces available to the gateway for an existing allocation
```
Comment (Anirudh9794) [inline pkg/vcdsdk/gateway.go:2228]: nit: releaseIpFromLoadBalancer -> ReleaseIpFromLoadBalancer

File: pkg/vcdsdk/gateway.go [hunk 2]
```diff
@@ -2209,3 +2224,51 @@ func (gm *GatewayManager) reserveIpForLoadBalancer(ctx context.Context, claimMar
...
+// (description of allocation will contain cluster id, service name and namespace). If such an allocation can't be
+// retrieved, the method will return without raising any error. It should be assumed that the allocation was removed in
+// a previous attempt. If an allocation is found, it be deleted, thereby releasing the IP from the load balancer as well
```
Comment (Anirudh9794) [inline pkg/vcdsdk/gateway.go:2231]: nit: "If an allocation is found, it be deleted" -> "If an allocation is found, it **should** be deleted"

**Review by Anirudh9794 — APPROVED** [review]
LGTM

**Review by rocknes — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/349)* status: all files reviewed, 2 unresolved discussions (waiting on @Anirudh9794, @arunmk, and @sahithi)

___
*[`pkg/vcdsdk/gateway.go` line 2228 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/349#gh-1488722417:-NqZxxnv83SUu2oXTXfS:bx3iq6m) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/43731dbe4021de52f07d60cdfde8362e33d24c80/pkg/vcdsdk/gateway.go#L2228)):*
<details><summary><i>Previously, Anirudh9794 wrote…</i></summary><blockquote>

nit: releaseIpFromLoadBalancer -> ReleaseIpFromLoadBalancer
</blockquote></details>

Addressed
___
*[`pkg/vcdsdk/gateway.go` line 2231 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/349#gh-1488723298:-NqZxykf-8Xmr-m0K6nj:bnna4xe) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/43731dbe4021de52f07d60cdfde8362e33d24c80/pkg/vcdsdk/gateway.go#L2231)):*
<details><summary><i>Previously, Anirudh9794 wrote…</i></summary><blockquote>

nit: "If an allocation is found, it be deleted" -> "If an allocation is found, it **should** be deleted"
</blockquote></details>

Thanks, Fixed

File: pkg/vcdsdk/gateway.go [hunk 3]
```diff
@@ -2209,3 +2224,51 @@ func (gm *GatewayManager) reserveIpForLoadBalancer(ctx context.Context, claimMar
...
+			updatedIpSpaceAllocation, err := gm.MarkIpAsUnused(ipSpaceAllocation)
+			if err != nil {
+				return fmt.Errorf("unable to mark IP [%s] from IP Space [%s] as unused. error [%v]", allocatedIp, ipSpace.IpSpace.Name, err)
+			}
+			err = gm.ReleaseIp(updatedIpSpaceAllocation)
+			if err != nil {
+				return fmt.Errorf("unable to release IP [%s] from IP Space [%s]. error [%v]", allocatedIp, ipSpace.IpSpace.Name, err)
+			}
+			// No need to process any more IP spaces, it is safe to return from this function
+			break
```
Comment (arunmk) [inline pkg/vcdsdk/gateway.go]: Maybe we should return here itself.

**Review by arunmk — APPROVED** [review]
LGTM. Please note that the Remove functionality will be called from CSE backend as well, during a force-delete. The interface should account for it.

**Pull request merged**